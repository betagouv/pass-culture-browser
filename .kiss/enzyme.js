import React from 'react'
import { shallow } from 'enzyme'

import MyComponent from '../MyComponent'

describe('src | components | MyComponent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const props = { errors: {} }
      const wrapper = shallow(<MyComponent {...props} />)
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
