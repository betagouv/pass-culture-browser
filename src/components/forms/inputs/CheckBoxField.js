/* eslint
react/jsx-one-expression-per-line: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import FormError from '../FormError'
import { isEmpty } from '../../../utils/strings'

const DEFAULT_REQUIRED_ERROR = 'Ce champs est requis'
const validateRequiredField = value => {
  if (value && !isEmpty(value)) return undefined
  return DEFAULT_REQUIRED_ERROR
}

const CheckBoxField = ({ children, className, name, required }) => {
  const validateFunc =
    required && typeof required === 'function'
      ? required
      : (required && validateRequiredField) || undefined

  return (
    <Field
      name={name}
      type="checkbox"
      validate={validateFunc || undefined}
      render={({ input, meta }) => (
        <p className={`${className}`}>
          <span className={className}>
            <input
              {...input}
              type="checkbox"
              className="input no-background"
              required={!!required}
            />
            {children}
            <FormError meta={meta} />
          </span>
        </p>
      )}
    />
  )
}

CheckBoxField.defaultProps = {
  className: null,
  required: false,
}

CheckBoxField.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
}

export default CheckBoxField
