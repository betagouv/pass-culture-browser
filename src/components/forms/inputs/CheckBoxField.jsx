import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import FormError from '../FormError'
import { isEmpty } from '../../../utils/strings'

const DEFAULT_REQUIRED_ERROR = 'Ce champ est requis'
const validateRequiredField = value => {
  if (value && !isEmpty(value)) return undefined
  return DEFAULT_REQUIRED_ERROR
}

class CheckBoxField extends Component {
  renderField = ({ input, meta }) => {
    const { children, className, name, required } = this.props

    return (
      <Fragment>
        <label
          className={`pc-final-form-contact ${className}`}
          htmlFor={name}
        >
          <input
            {...input}
            className="input no-background"
            id={name}
            required={required}
            type="checkbox"
          />
          {children}
        </label>
        <FormError meta={meta} />
      </Fragment>
    )
  }

  render() {
    const { name, required } = this.props
    const validateFunc =
      required && typeof required === 'function'
        ? required
        : (required && validateRequiredField) || undefined

    return (
      <Field
        name={name}
        render={this.renderField}
        type="checkbox"
        validate={validateFunc || undefined}
      />
    )
  }
}

CheckBoxField.defaultProps = {
  className: '',
  required: true,
}

CheckBoxField.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  required: PropTypes.bool,
}

export default CheckBoxField
