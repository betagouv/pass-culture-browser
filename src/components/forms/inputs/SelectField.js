import React from 'react'
import PropTypes from 'prop-types'
import { Form, Select } from 'antd'
import { Field } from 'react-final-form'

import { renderLabel } from '../utils'

const filterOption = (input, option) => {
  const child = option.props.children
  return child.toLowerCase().indexOf(input.toLowerCase()) >= 0
}

class SelectField extends React.PureComponent {
  constructor(props) {
    super(props)
    this.popupContainer = null
  }

  setContainerRef = ref => {
    this.popupContainer = ref
  }

  renderField = ({ input }) => {
    const { canSearch, disabled, placeholder, provider, readOnly } = this.props

    const isdisabled = readOnly || disabled || !provider.length || provider.length === 1
    const moreprops = { disabled: isdisabled }

    return (
      <Select
        {...moreprops}
        filterOption={filterOption}
        getPopupContainer={this.getPopupContainer()}
        onChange={input.onChange}
        optionFilterProp="children"
        placeholder={placeholder}
        showSearch={canSearch}
        size="large"
        value={input.value || undefined}
      >
        {provider &&
          provider.map(obj => (
            <Select.Option
              key={obj.id}
              value={obj.id}
            >
              {obj.label}
            </Select.Option>
          ))}
      </Select>
    )
  }

  getPopupContainer = () => (
    this.popupContainer
  )

  render() {
    const { label, help, id, name, disabled, provider, className, readOnly, ...rest } = this.props
    const isdisabled = readOnly || disabled || !provider.length || provider.length === 1

    const moreprops = { disabled: isdisabled }
    if (id) moreprops.id = id
    return (
      <Form.Item
        {...rest}
        className={`ant-select-custom ${className}`}
        label={(label && renderLabel(label, help)) || null}
      >
        <Field
          name={name}
          render={this.renderField}
        />
        <div
          className="select-field-popup-container is-relative"
          ref={this.setContainerRef}
        />
      </Form.Item>
    )
  }
}

SelectField.defaultProps = {
  canSearch: false,
  className: '',
  disabled: false,
  help: null,
  id: null,
  label: null,
  placeholder: null,
  readOnly: false,
}

SelectField.propTypes = {
  canSearch: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  help: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  provider: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  readOnly: PropTypes.bool,
}

export default SelectField
