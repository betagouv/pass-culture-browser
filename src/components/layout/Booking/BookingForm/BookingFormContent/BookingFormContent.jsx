import classnames from 'classnames'
import PropTypes from 'prop-types'
import React, { Component, Fragment } from 'react'
import { Field } from 'react-final-form'

import { getCalendarProvider, parseHoursByStockId } from '../../utils'
import SelectField from '../../../../forms/inputs/SelectField'
import DatePickerField from '../../../../forms/inputs/DatePickerField/DatePickerField'

class BookingFormContent extends Component {
  componentDidMount() {
    const { invalid, onChange, values } = this.props
    const { price, stockId } = values
    if (stockId) {
      const nextCanSubmitForm = Boolean(!invalid && stockId && price >= 0)
      onChange(nextCanSubmitForm)
    }
  }

  componentDidUpdate(prevProps) {
    const { invalid, onChange, values } = this.props
    const { price, stockId } = values
    if (stockId !== prevProps.values.stockId) {
      const nextCanSubmitForm = Boolean(!invalid && stockId && price >= 0)
      onChange(nextCanSubmitForm)
    }
  }

  handleChangeAndRemoveCalendar = input => (value, event) => {
    event.preventDefault()
    input.onChange(value)
  }

  renderBookingDatePickerField = ({ input }) => {
    const { values } = this.props
    const { value } = input
    const calendarDates = getCalendarProvider(values)
    const calendarLabel = calendarDates.length === 1 ? '' : 'Choisissez une date'
    const dateFormat = 'DD MMMM YYYY'
    return (
      <DatePickerField
        {...input}
        className="text-center mb36"
        clearable={false}
        dateFormat={dateFormat}
        hideToday
        id="booking-form-date-picker-field"
        includeDates={calendarDates}
        label={calendarLabel}
        name="date"
        onChange={this.handleChangeAndRemoveCalendar(input)}
        popperPlacement="bottom"
        readOnly={calendarDates.length === 1}
        selected={value}
      />
    )
  }

  render() {
    const { className, formId, handleSubmit, isEvent, isReadOnly, values } = this.props
    const { price } = values
    const bookableTimes = parseHoursByStockId(values)
    const hasOneBookableTime = bookableTimes.length === 1
    const hourLabel = hasOneBookableTime ? '' : 'Choisissez une heure'
    return (
      <form
        className={classnames(className, {
          'is-read-only': isReadOnly,
        })}
        id={formId}
        onSubmit={handleSubmit}
      >
        {isEvent && (
          <Fragment>
            <Field
              name="date"
              render={this.renderBookingDatePickerField}
            />
            {bookableTimes && (
              <SelectField
                className="text-center"
                id="booking-form-time-picker-field"
                label={hourLabel}
                name="time"
                options={bookableTimes}
                placeholder="Heure et prix"
                readOnly={hasOneBookableTime}
              />
            )}
          </Fragment>
        )}

        {!isEvent && (
          <p className="text-center fs22">
            <span className="is-block">{'Vous êtes sur le point de réserver'}</span>
            <span className="is-block">{`cette offre pour ${price} €.`}</span>
          </p>
        )}
      </form>
    )
  }
}

BookingFormContent.defaultProps = {
  className: '',
  isEvent: false,
  isReadOnly: false,
}

BookingFormContent.propTypes = {
  className: PropTypes.string,
  formId: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  isEvent: PropTypes.bool,
  isReadOnly: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  values: PropTypes.shape().isRequired,
}

export default BookingFormContent
