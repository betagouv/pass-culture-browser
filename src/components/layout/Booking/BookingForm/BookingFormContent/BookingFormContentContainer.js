import { connect } from 'react-redux'

import BookingFormContent from './BookingFormContent'
import { selectIsStockDuo } from '../../../../../redux/selectors/data/stocksSelectors'
import { selectOfferById } from '../../../../../redux/selectors/data/offersSelectors'

export const mapStateToProps = (state, ownProps) => {
  const { offerId, values } = ownProps
  const { stockId } = values
  const isStockDuo = selectIsStockDuo(state, stockId, offerId)
  const offer = selectOfferById(state, offerId)
  const canExpire = offer.offerType.canExpire
  const isDigital = offer.isDigital

  return {
    isDigital,
    isStockDuo,
    canExpire,
  }
}

export default connect(mapStateToProps)(BookingFormContent)
