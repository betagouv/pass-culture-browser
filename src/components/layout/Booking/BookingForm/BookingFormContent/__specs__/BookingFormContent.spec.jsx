import { shallow, mount } from 'enzyme'
import moment from 'moment'
import React from 'react'
import { Field } from 'react-final-form'

import CheckBoxField from '../../../../../forms/inputs/CheckBoxField'
import SelectField from '../../../../../forms/inputs/SelectField'
import BookingFormContent from '../BookingFormContent'

describe('bookingFormContent', () => {
  let props
  let handleSubmit

  beforeEach(() => {
    handleSubmit = jest.fn()
    props = {
      extraClassName: 'fake className',
      formId: 'fake formId',
      invalid: false,
      isStockDuo: false,
      isEvent: false,
      isReadOnly: false,
      handleSubmit,
      onChange: jest.fn(),
      offerId: 'o1',
      values: {
        bookables: [],
        date: null,
        price: null,
      },
    }
  })

  it('should render a form element with the proper props when is not read only mode', () => {
    // when
    const wrapper = mount(<BookingFormContent {...props} />)

    // then
    const form = wrapper.find('form')
    expect(form).toHaveLength(1)
    expect(form.prop('className')).toBe('fake className ')
    expect(form.prop('id')).toBe('fake formId')
    expect(form.prop('onSubmit')).toStrictEqual(handleSubmit)
  })

  it('should render a form element with the proper props when is read only mode', () => {
    // given
    props.isReadOnly = true

    // when
    const wrapper = mount(<BookingFormContent {...props} />)

    // then
    const form = wrapper.find('form')
    expect(form).toHaveLength(1)
    expect(form.prop('className')).toBe('fake className is-read-only')
    expect(form.prop('id')).toBe('fake formId')
    expect(form.prop('onSubmit')).toStrictEqual(handleSubmit)
  })

  describe('when booking a Duo offer', () => {
    it('should display a field and checkbox', () => {
      // given
      props.isStockDuo = true
      props.isEvent = true
      props.values = {
        bookables: [
          {
            id: 'B1',
          },
        ],
        date: '21/10/2001',
        price: 5,
      }

      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      const field = wrapper.find(Field)
      const checkBoxField = wrapper.find(CheckBoxField)
      expect(field).toHaveLength(1)
      expect(checkBoxField).toHaveLength(1)
    })
  })

  describe('when booking an event', () => {
    beforeEach(() => {
      props.isEvent = true
    })

    it('should display a notice regarding the withdrawal period', () => {
      // given
      props.isEvent = true
      props.values = {
        bookables: [
          {
            id: 'B1',
          },
        ],
        date: '21/10/2001',
        price: 5,
      }

      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      expect(wrapper.find(
        { children: 'Réservation annulable dans les 48h suivants la réservation et moins de 72h avant la date de l’évènement (si applicable)' }
      )).toHaveLength(1)
    })

    it('should render a Field component with the proper props', () => {
      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      const field = wrapper.find(Field)
      expect(field).toHaveLength(1)
      expect(field.prop('name')).toBe('date')
      expect(field.prop('render')).toStrictEqual(expect.any(Function))
    })

    it('should not render a SelectField component when no date has been selected', () => {
      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      const selectField = wrapper.find(SelectField)
      expect(selectField).toHaveLength(0)
    })

    it('should display the selected date if one has been selected', () => {
      // given
      const date = moment().add(3, 'days')
      props.values = {
        bookables: [
          {
            beginningDatetime: date,
            id: 'AE',
            price: 12,
          },
          {
            beginningDatetime: date,
            id: 'AF',
            price: 13,
          },
        ],
        date,
      }

      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      const selectField = wrapper.find(SelectField)
      expect(selectField).toHaveLength(1)
      expect(selectField.props()).toStrictEqual({
        className: 'text-center',
        disabled: false,
        id: 'booking-form-time-picker-field',
        label: 'Choisis une heure :',
        name: 'time',
        placeholder: 'Heure et prix',
        options: [
          {
            id: 'AE',
            label: `${date.format('HH:mm')} - 12\u00A0€`,
          },
          {
            id: 'AF',
            label: `${date.format('HH:mm')} - 13\u00A0€`,
          },
        ],
        readOnly: false,
        required: false,
      })
    })
  })

  describe('when booking a thing', () => {
    beforeEach(() => {
      props.isEvent = false
    })
    it('should display the corresponding messages', () => {
      // given
      props.values = {
        bookables: [],
        date: '2019-01-01',
        price: 12,
      }

      // when
      const wrapper = shallow(<BookingFormContent {...props} />)

      // then
      expect(wrapper.find({ children: 'Tu es sur le point de réserver' })).toHaveLength(1)
      expect(wrapper.find({ children: 'cette offre pour 12 €.' })).toHaveLength(1)
    })
  })
})
