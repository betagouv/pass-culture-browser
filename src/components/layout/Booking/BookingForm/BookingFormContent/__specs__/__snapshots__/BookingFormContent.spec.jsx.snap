// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | Booking | BookingForm | BookingFormContent should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookingFormContent
    className="fake className"
    formId="fake formId"
    handleSubmit={[MockFunction]}
    invalid={false}
    isEvent={false}
    isReadOnly={false}
    onChange={[MockFunction]}
    values={
      Object {
        "bookables": Array [],
        "date": "2019-01-01",
        "price": 12,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <p
          className="text-center fs22"
        >
          <span
            className="is-block"
          >
            Vous êtes sur le point de réserver
          </span>
          <span
            className="is-block"
          >
            cette offre pour 12 €.
          </span>
        </p>,
      ],
      "className": "fake className",
      "id": "fake formId",
      "onSubmit": [MockFunction],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="is-block"
            >
              Vous êtes sur le point de réserver
            </span>,
            <span
              className="is-block"
            >
              cette offre pour 12 €.
            </span>,
          ],
          "className": "text-center fs22",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Vous êtes sur le point de réserver",
              "className": "is-block",
            },
            "ref": null,
            "rendered": "Vous êtes sur le point de réserver",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "cette offre pour 12 €.",
              "className": "is-block",
            },
            "ref": null,
            "rendered": "cette offre pour 12 €.",
            "type": "span",
          },
        ],
        "type": "p",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <p
            className="text-center fs22"
          >
            <span
              className="is-block"
            >
              Vous êtes sur le point de réserver
            </span>
            <span
              className="is-block"
            >
              cette offre pour 12 €.
            </span>
          </p>,
        ],
        "className": "fake className",
        "id": "fake formId",
        "onSubmit": [MockFunction],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="is-block"
              >
                Vous êtes sur le point de réserver
              </span>,
              <span
                className="is-block"
              >
                cette offre pour 12 €.
              </span>,
            ],
            "className": "text-center fs22",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Vous êtes sur le point de réserver",
                "className": "is-block",
              },
              "ref": null,
              "rendered": "Vous êtes sur le point de réserver",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "cette offre pour 12 €.",
                "className": "is-block",
              },
              "ref": null,
              "rendered": "cette offre pour 12 €.",
              "type": "span",
            },
          ],
          "type": "p",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
